src/rsa/signing.rs

```diff
@@ -125,11 +125,34 @@ impl RSAKeyPair {
                 if p.bit_length() != q.bit_length() {
                     return Err(error::Unspecified);
                 }
+                let q = try!(q.into_odd_positive());
+                let n = try!(n.into_modulus::<N>());
+
+                let q_mod_n_decoded = {
+                    let q = try!(q.try_clone());
+                    try!(q.into_elem_decoded(&n))
+                };
+
                 // XXX: |p < q| is actual OK, it seems, but our implementation
                 // of CRT-based moduluar exponentiation used requires that
                 // |q > p|. (|p == q| is just wrong.)
-                let q = try!(q.into_odd_positive());
                 try!(bigint::verify_less_than(&q, &p));
+                {
+                    let p_mod_n = {
+                        let p = try!(p.try_clone());
+                        try!(p.into_elem_decoded(&n))
+                    };
+                    let p_minus_q = {
+                        let p_minus_q = try!(bigint::elem_sub(
+                                p_mod_n, &q_mod_n_decoded, &n));
+                        try!(p_minus_q.into_positive())
+                    };
+                    let min_pq_bitlen_diff = try!(half_n_bits.try_sub(
+                            bits::BitLength::from_usize_bits(100)));
+                    if p_minus_q.bit_length() < min_pq_bitlen_diff {
+                        return Err(error::Unspecified);
+                    }
```

Comment:
Sorry, I could very well be wrong, but I think there's one small (and probably insignificant) edge case. Let's consider a simpler analogous problem: Let's say we want to accept x if and only if `lg x > 2`, i.e. reject x when `lg x ≤ 2`:

| `x` | `lg x`  | `⌊lg x⌋` |
|---|-------|--------|
| 1 | 0     | 0      |
| 2 | 1     | 1      |
| 3 | 1.58  | 1      |
| 4 | 2     | 2      |
| 5 | 2.32  | 2      |

Notice that in one case where `⌊lg x⌋ = 2` (`x = 4`, i.e. the case where `⌊lg x⌋ == lg x`), we should reject `x`, but in another case where `⌊lg x⌋ = 2` (`x = 5`) we accept `x`. IIUC, there's no way to handle this edge case perfectly correctly given only `⌊lg x⌋` but not whether or not `⌊lg x⌋ == lg x`, right?

Going back to our actual problem, IMO it doesn't matter what we do in the case where p - q is exactly equal to 2<sup>nBits/2 - 100</sup>. The important thing is that we minimize the amount of code we add for this check and that we document clearly something along the lines of "This code accepts p - q == 2**(nBits/2 - 100) even though the spec says that value is the largest value that should be rejected, because it's too much work to handle that one value the way the spec says." (I hope you can come up with better wording than I did.)

(See http://crypto.stackexchange.com/a/35096 for background.)